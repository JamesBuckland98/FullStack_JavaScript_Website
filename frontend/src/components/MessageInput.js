import Chat from './Chat';
import React, { Component } from 'react';
import getWeather from '../api/weatherApi';
import { saveChatHistory, sendMessage } from '../api/chatApi';
import { geolocated } from "react-geolocated";
import { Formik, Form, Field, ErrorMessage, } from 'formik';

class MessageInput extends Component {

	constructor(props) {
		super(props)
		this.state = {
			weather : 0,
		}
	}

	componentDidUpdate(prevProps) {
		// https://reactjs.org/docs/react-component.html#shouldcomponentupdate
		// wrapped in a condition to stop infinite call to API
		if (this.props.coords !== prevProps.coords) {
			getWeather( this.props.coords.longitude ,  this.props.coords.latitude , (err, response) => {
				this.setState({
					weather: response.main.temp, 
				})
			});
		}
	}
	
	render() {
		return (
			<div>
				<p> 
					Welcome, { this.props.username } <br/>
					{ !this.props.isGeolocationAvailable ? <span> Your browser does not support Geolocation </span> : null }
				</p>

				<Chat 
					weather = { this.state.weather } 
					messages = { this.props.messages }
					username = { this.props.username }
				/>

				<Formik
				initialValues={{ChatMessage: '', Species: '', Abundance: '' }}
				validate= {
					values => {
						let errors = {};
						if ( values.file && values.file.type !== 'image/png') {
							errors.file = 'File must be a .png or .jpg'
						}
						if ( ! values.ChatMessage) {
							errors.ChatMessage = 'Required'
						}
						if( ! values.Species ) {
							errors.Species = 'Required'
						}
						if( ! values.Abundance) {
							errors.Abundance = ' Required'
						}
						return errors;
					}
				}

				onSubmit = {(values, { setSubmitting, resetForm }) => {
					setTimeout(() => {
						// Adds addition information before passing the value
						Object.assign(values, { 
							ChatTimeStamp : new Date().toISOString().slice(0, 19).replace('T', ' '),
							Username: this.props.username,
							Longitude: this.props.coords.longitude,
							Latitude: this.props.coords.latitude,
							Weather: this.state.weather,
						})
						setSubmitting(false);
						saveChatHistory(values);
						sendMessage(values);
					}, 400);
					// Resets to initial values
					resetForm({});
				}}
				>

				{({ isSubmitting, setFieldValue }) => (
					<Form className="main">
						<Field type="text" name="ChatMessage" placeholder = "Enter Title"></Field>
						<ErrorMessage name="ChatMessage" component="div" /> 
						<Field type="text" name="Species" placeholder="Enter Species"/>
						<ErrorMessage name="Species" component="div" /> 
						<Field type= "number" name="Abundance" placeholder="Enter Abundance"/>
						<ErrorMessage name="Abundance" component="div" /> 
						<input id="file" name="file" type="file" onChange={
							(event) => {
								setFieldValue("file", event.currentTarget.files[0]);
							}
						} className="form-control" />
						<ErrorMessage name="file" component="div" />
						<button type="submit" disabled={isSubmitting}>
						Submit
						</button>
					</Form>
				)}
				</Formik>
			</div>
		)
	}
}
export default geolocated({
    positionOptions: {
        enableHighAccuracy: false,
    },
    userDecisionTimeout: null,
})(MessageInput);