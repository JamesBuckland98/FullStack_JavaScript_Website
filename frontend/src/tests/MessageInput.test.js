import React from 'react';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
configure({ adapter: new Adapter() });
import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';
import MessageInput from '../components/MessageInput';
import { geolocated } from "react-geolocated";

it('should render Message input component correctly', () => {
	const mockSuccessfulGeolocationProvider = {
		getCurrentPosition(onSuccess) {
			return onSuccess({
				coords: {
					latitude: 50,
					longitude: 20,
				},
			});
		},
		watchPosition(onSuccess) {
			return onSuccess({
				coords: {
					latitude: 50,
					longitude: 20,
				},
			});
		},
	};
	const Component = geolocated({
		geolocationProvider: mockSuccessfulGeolocationProvider,
	})(MessageInput)
	const wrapper = shallow(<Component/>)
	expect(toJson(wrapper)).toMatchSnapshot();
});

