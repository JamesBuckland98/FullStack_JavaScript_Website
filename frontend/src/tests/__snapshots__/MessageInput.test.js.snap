// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Message input component correctly 1`] = `
<Geolocated(MessageInput)
  coords={
    Object {
      "latitude": 50,
      "longitude": 20,
    }
  }
  isGeolocationAvailable={true}
  isGeolocationEnabled={true}
  positionError={null}
/>
`;
